brave/brave-core
# Update Brave News to use the new `channels` field from the API (when available)

Resolves https://github.com/brave/brave-browser/issues/25688

Production sources:
https://brave-today-cdn.brave.com/sources.json

Staging sources:
https://brave-today-cdn.bravesoftware.com/sources.global.json

This PR is in the somewhat awkward position of adding new functionality from a new, future version of the sources API. The PR adds the ability to parse the `channel` field if it exists, and when it does not, falls back to the existing behavior.

This PR does the following:
1. Migrate from using `publisher.isEnabled` to using the `Top Sources` channel to decide which sources are available by default (on Desktop) if (and only if) there are `channels` available.  
2. Adds a new switch `--brave-news-region` to override the current region
3. Adds a `ChannelsController` for fetching `Channels` and managing channel subscriptions
4. Updates `BuildFeed` to take Channel subscriptions into account.
5. Subscribes all users to `Top Sources` channel by default (this has no effect if the server doesn't return channels).

## Running the PR
To use the new channels before the API change lands in prod, you need to add the following two flags to point at the staging `sources.json`.

`npm run start -- --brave-today-host=brave-today-cdn.bravesoftware.com --brave-news-region=en_US`

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

1. Enable Brave News
3. Articles should be present in the feed and a number of categories and feeds should be present.
4. Enable the brave://flags/#brave-news-v2 flag
5. Close Brave
6. Relaunch with the command line argument `--brave-today-host=brave-today-cdn.bravesoftware.com`
7. The feed `Gamespot` should be present in the `Entertainment`, `Gaming` and `Top Sources` categories.
8. Articles should still show up in the feed.
